File: hydro/src/main/java/com/ultreon/hydro/Game.java
package com.ultreon.hydro;

import com.ultreon.commons.crash.CrashLog;
import com.ultreon.commons.crash.ApplicationCrash;
import com.ultreon.commons.lang.InfoTransporter;
import com.ultreon.commons.lang.LoggableProgress;
import com.ultreon.commons.time.TimeProcessor;
import com.ultreon.hydro.event.RenderGameEvent;
import com.ultreon.hydro.event.RenderScreenEvent;
import com.ultreon.hydro.player.IPlayer;
import com.ultreon.hydro.player.PlayerController;
import com.ultreon.hydro.event.TickEvent;
import com.ultreon.hydro.event.bus.GameEvents;
import com.ultreon.hydro.render.FilterApplier;
import com.ultreon.hydro.render.RenderSettings;
import com.ultreon.hydro.render.Renderer;
import com.ultreon.hydro.resources.ResourceManager;
import com.ultreon.hydro.screen.Screen;
import com.ultreon.hydro.screen.ScreenManager;
import lombok.SneakyThrows;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.awt.image.BufferStrategy;
import java.awt.image.BufferedImageOp;
import java.awt.image.ImageObserver;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

public abstract class Game {
    static Game instance;

    // Values
    private int fps;
    private final int tps;

    // Modes
    private static boolean debugMode;
    private static boolean devMode;

    // Utility objects.
    public IPlayer playerInterface;
    private PlayerController playerController;
    protected final ResourceManager resourceManager;
    protected final GameWindow gameWindow;
    protected final ScreenManager screenManager;
    protected final RenderSettings renderSettings;

    // Game states.
    private boolean loaded;
    private boolean crashed;
    private volatile boolean running = false;

    // Tasks
    private final List<Runnable> tasks = new CopyOnWriteArrayList<>();

    // Threads
    protected Thread mainThread;

    public static Game getInstance() {
        return instance;
    }

    public Game(GameWindow.Properties windowProperties, BootOptions bootOptions) {
        // Set default uncaught exception handler.
        Thread.setDefaultUncaughtExceptionHandler(new GameExceptionHandler(this));

        // Assign instance.
        instance = this;

        // Set game properties.
        this.tps = bootOptions.tps;

        // Prepare for game launch
        this.resourceManager = new ResourceManager();
        this.renderSettings = new RenderSettings();

        // Setup game window.
        this.gameWindow = new GameWindow(windowProperties);

        // Prepare for loading.
        this.prepare();

//        this.gameWindow.init();

        // Load game with loading screen.
        this.load(new LoggableProgress(new InfoTransporter(this::log), 1000));
        this.screenManager = createScreenManager();
    }

    public final void scheduleTask(Runnable task) {
        this.tasks.add(task);
    }

    /**
     * Initialize the game engine.
     *
     * @param debugMode debug mode enables debugging utilities.
     * @param devMode dev mode enabled development utilities.
     */
    public static void initEngine(boolean debugMode, boolean devMode) {
        Game.debugMode = debugMode;
        Game.devMode = devMode;
    }

    private void log(String text) {
        // Todo: implement
    }

    /**
     * This method should be overridden when filters are used in the game.
     *
     * @return list of buffered image operations.
     * @implNote should be overridden for enabling / disabling render filters.
     */
    @NotNull
    public List<BufferedImageOp> getCurrentFilters() {
        return new ArrayList<>();
    }

    public abstract ScreenManager createScreenManager();

    protected abstract int getMainLoadingSteps();

    /**
     * Todo: allow for adding multiple loading progresses.
     *
     * @implNote should always be overridden as protected.
     * @param mainProgress main loading progress.
     */
    protected abstract void load(LoggableProgress mainProgress);

    protected abstract void prepare();

    protected abstract void preparePlayer();

    protected abstract IPlayer createPlayer();

    protected abstract void render(Renderer renderer);

    protected abstract void tick();

    protected abstract void loadEnvironment();

    public void loadPlayEnvironment() {
        this.playerInterface = createPlayer();
        this.playerController = new PlayerController(this.playerInterface);

        loadEnvironment();
    }

    /////////////////////
    //     Getters     //
    /////////////////////
    public boolean isAntialiasEnabled() {
        return true;
    }

    public boolean isTextAntialiasEnabled() {
        return isAntialiasEnabled();
    }

    public final GameWindow getGameWindow() {
        return this.gameWindow;
    }

    public ImageObserver getObserver() {
        return gameWindow.observer;
    }

    public Rectangle getBounds() {
        return new Rectangle(0, 0, gameWindow.getWidth(), gameWindow.getHeight());
    }

    public final int getWidth() {
        return gameWindow.getWidth();
    }

    public final int getHeight() {
        return gameWindow.getHeight();
    }

    public int getFps() {
        return fps;
    }

    public Cursor getDefaultCursor() {
        return Cursor.getDefaultCursor();
    }

    public abstract void onClose();

    public ScreenManager getScreenManager() {
        return screenManager;
    }

    public static boolean isDebugMode() {
        return Game.debugMode;
    }

    public static boolean isDevMode() {
        return Game.devMode;
    }

    public boolean isLoaded() {
        return this.loaded;
    }

    protected void markLoaded() {
        this.loaded = true;
    }

    public ResourceManager getResourceManager() {
        return resourceManager;
    }

    public RenderSettings getRenderSettings() {
        return renderSettings;
    }

    private void mainThread() {
        double tickCap = 1d / (double) tps;
        double frameTime = 0d;
        double frames = 0;

        double time = TimeProcessor.getTime();
        double unprocessed = 0;

        try {
            while (running) {
                boolean canTick = false;

                double time2 = TimeProcessor.getTime();
                double passed = time2 - time;
                unprocessed += passed;
                frameTime += passed;

                time = time2;

                while (unprocessed >= tickCap) {
                    unprocessed -= tickCap;

                    canTick = true;
                }

                if (canTick) {
                    try {
                        internalTick();
                    } catch (Throwable t) {
                        CrashLog crashLog = new CrashLog("Game being ticked.", t);
                        throw crashLog.createCrash();
                    }
                }

                if (frameTime >= 1.0d) {
                    frameTime = 0;
                    fps = (int) Math.round(frames);
                    frames = 0;
                }

                frames++;

                try {
                    filteredRender(fps);
                } catch (Throwable t) {
                    CrashLog crashLog = new CrashLog("Game being rendered.", t);
                    throw crashLog.createCrash();
                }

                for (Runnable task : tasks) {
                    task.run();
                }

                tasks.clear();
            }
        } catch (ApplicationCrash e) {
            e.printStackTrace();

            Game.instance.crash(e);
//            showScreen(new CrashScreen(e.getCrashReport()), true);
            while (running) Thread.onSpinWait();
        }

        close();
    }

    /**
     *
     */
    private void internalTick() {
        @Nullable Screen currentScreen = screenManager.getCurrentScreen();

        if (playerController != null) {
            playerController.tick();
        }

        this.tick();

        // Call tick event.
        if (isLoaded() && (currentScreen == null || !currentScreen.doesPauseGame())) {
            GameEvents.get().publish(new TickEvent(this));
        }
    }

    /**
     * Render method, for rendering window.
     * @param fps current game framerate.
     */
    private void filteredRender(int fps) {
        if (!gameWindow.isInitialized()) return;

        // Buffer strategy (triple buffering).
        BufferStrategy bs = this.gameWindow.canvas.getBufferStrategy();

        // Create buffers if not created yet.
        if (bs == null) {
            this.gameWindow.canvas.createBufferStrategy(2);
            bs = this.gameWindow.canvas.getBufferStrategy();
        }

        // Get GraphicsProcessor and GraphicsProcessor objects.
        Renderer renderer = new Renderer(bs.getDrawGraphics());

        FilterApplier filterApplier = new FilterApplier(getBounds().getSize(), this.getObserver());
        Renderer filterRenderer = filterApplier.getRenderer();

        if (this.renderSettings.isAntialiasingEnabled() && this.isTextAntialiasEnabled())
            filterRenderer.hint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON);
        if (this.renderSettings.isAntialiasingEnabled() && this.isAntialiasEnabled())
            filterRenderer.hint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);

        List<BufferedImageOp> filters =  Game.instance.getCurrentFilters();

        this.filteredRender(filterRenderer);

        // Set filter gotten from filter event-handlers.
        filterApplier.setFilters(filters);

        // Draw filtered image.
        renderer.image(filterApplier.done(), 0, 0);

        // Disable Antialias
        // Todo: check performance.
//        if (renderSettings.isAntialiasingEnabled() && GameSettings.instance().isTextAntialiasEnabled())
//            renderer.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_OFF);
//        if (renderSettings.isAntialiasingEnabled() && GameSettings.instance().isAntialiasEnabled())
//            renderer.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_OFF);

        this.fps = fps;

        // Dispose and show.
        renderer.dispose();

        try {
            bs.show();
        } catch (IllegalStateException e) {
//            close();
        }
    }

    /**
     *
     * @param renderer the renderer to render the game with.
     */
    private void filteredRender(Renderer renderer) {

        // Call to game environment rendering.
        GameEvents.get().publish(new RenderGameEvent.Before(renderer));
        render(renderer);
        GameEvents.get().publish(new RenderGameEvent.After(renderer));

        // Get screen.
        @Nullable Screen screen = this.screenManager.getCurrentScreen();

        // Null check for screen variable.
        if (screen != null) {

            // Render the screen.
            GameEvents.get().publish(new RenderScreenEvent.Before(screen, renderer));
            screen.render(this, renderer);
            screen.renderGUI(this, renderer);
            GameEvents.get().publish(new RenderScreenEvent.After(screen, renderer));
        }
    }

    @SneakyThrows
    public void shutdown() {
        this.running = false;
        mainThread.join();
        close();
    }

    protected final void close() {
        onClose();
        gameWindow.close();
        if (crashed) {
            System.exit(0);
        }
    }

    /**
     * Handler for game crash.
     *
     * @param crash the game crash.
     */
    public void crash(ApplicationCrash crash) {
        CrashLog crashLog = crash.getCrashLog();
        crashed = true;

        boolean overridden = false;
        try {
            overridden = onCrash(crash);
        } catch (Throwable ignored) {

        }

        if (!overridden) {
            crashLog.defaultSave();
            crashLog.createCrash().printCrash();
        }
    }

    /**
     * Crash handler, for overriding default crash handling.
     *
     * @param crash the crash that happened.
     * @return whether the default should be overridden or not. The value {@code }
     */
    @SuppressWarnings("unused")
    public boolean onCrash(ApplicationCrash crash) {
        return false;
    }

    @SuppressWarnings("Convert2Lambda")
    void windowLoaded() {
        this.mainThread = new Thread(new Runnable() {
            public void run() {
                running = true;
                Game.this.mainThread();
            }
        }, "Main-Thread");
        this.mainThread.start();
    }

    public boolean isMainAlive() {
        return mainThread.isAlive();
    }

    public Cursor getPointerCursor() {
        return Cursor.getPredefinedCursor(Cursor.HAND_CURSOR);
    }

    protected static class BootOptions {
        private int tps = 20;

        public BootOptions() {

        }

        public BootOptions tps(int tps) {
            this.tps = tps;
            return this;
        }
    }
}

File: src/main/kotlin/qtech/hydro/Game.kt
package qtech.hydro

import org.apache.commons.lang3.SystemUtils
import org.apache.logging.log4j.Level
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import org.fusesource.jansi.AnsiConsole
import org.jdesktop.swingx.graphics.FilterComposite
import qtech.bubbles.BubbleBlaster
import qtech.bubbles.common.References
import qtech.bubbles.common.Timer
import qtech.bubbles.common.renderer.CursorRenderer
import qtech.hydro.crash.CrashExceptionHandler
import qtech.hydro.crash.CrashReport
import qtech.bubbles.common.scene.ScreenManager
import qtech.bubbles.common.screen.Screen
import qtech.bubbles.common.streams.CustomOutputStream
import qtech.bubbles.core.controllers.KeyboardController
import qtech.bubbles.core.controllers.MouseController
import qtech.bubbles.core.utils.categories.FileUtils
import qtech.bubbles.core.utils.categories.StringUtils
import qtech.bubbles.event.ExitEvent
import qtech.bubbles.event.FilterEvent
import qtech.bubbles.event.bus.BBEventBus
import qtech.bubbles.gui.Window
import qtech.bubbles.screen.CrashScreen
import qtech.bubbles.settings.GameSettings
import qtech.hydro.crash.CrashCategory
import qtech.hydro.crash.ReportedException
import qtech.hydro.ui.CursorManager
import qtech.hydro.ui.FontManager
import java.awt.*
import java.awt.image.BufferStrategy
import java.awt.image.ImageObserver
import java.io.File
import java.io.PrintStream
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList
import kotlin.math.roundToLong
import kotlin.system.exitProcess

abstract class Game(gameDir: File, val args: Array<String>) {
    var cursor: Cursor
        get() = window.cursor
        set(value) {
            window.cursor = value
        }
    val graphicsConfiguration: GraphicsConfiguration by lazy { canvas.graphicsConfiguration }
    val bounds: Rectangle
        get() = canvas.bounds
    var workDir: File = gameDir
        private set
    abstract val fullscreen: Boolean
    protected lateinit var window: Window

    lateinit var mainThread: Thread
        private set

    private val tasks: MutableList<Runnable> = CopyOnWriteArrayList()

    val eventBus = BBEventBus()
    var graphicsConfigurator: GraphicsConfigurator
        protected set
    @Volatile
    var running: Boolean = false
        protected set
    var screenManager: ScreenManager
        protected set

    open val defaultFont: Font
        get() = canvas.font
    open val defaultCursor: Cursor
        get() = Cursor.getDefaultCursor()

    protected val canvas: Canvas = Canvas()


    val currentScreen: Screen?
        get() = screenManager.currentScreen
    val isOnMainThread: Boolean
        get() = Thread.currentThread() === mainThread
    val fontName: String
        get() = canvas.font.fontName
    val font: Font
        get() = canvas.font
    var width: Int
        get() = canvas.width
        set(value) {
            window.setSize(value, window.height)
            canvas.setSize(value, window.height)
        }
    var height: Int
        get() = canvas.height
        set(value) {
            window.setSize(window.width, value)
            canvas.setSize(window.width, value)
        }

    val imageObserver: ImageObserver = ProtectedImageObserver(canvas)

    init {
        instance = this

        workDir = gameDir

        requireNotNull(workDir)

        val exceptionHandler = CrashExceptionHandler()
        Thread.setDefaultUncaughtExceptionHandler(exceptionHandler)

        try {
            screenManager = ScreenManager.instance
            initManagers()
            for (thread in Thread.getAllStackTraces().keys) {
                thread.uncaughtExceptionHandler = exceptionHandler
                if (thread.name == "JavaFX Application Thread") {
                    thread.name = "BubbleBlasterApp"
                }
                if (thread.name == "AWT-EventQueue-0") {
                    thread.name = "BB-EventBus"
                }
            }

            // Add ansi color compatibility in console.
            AnsiConsole.systemInstall()

            FileUtils.cwd = workDir

            val cursorManager = CursorManager()
            try {
                loadCursors(cursorManager)
            } catch (t: Throwable) {
                val report = CrashReport(t)
                val currentContext = cursorManager.currentContext
                if (currentContext != null) {
                    report.addCategory(CrashCategory("Cursor being loaded").also {
                        it.add("Name") { currentContext.name }
                        it.add("Image Class") { currentContext.image::class.qualifiedName }
                    })
                }
                val createContext = CursorRenderer.createContext
                if (createContext != null) {
                    report.addCategory(CrashCategory("Cursor being created").also {
                        it.add("Name") { createContext.name }
                    })
                }
                throw report.reportedException
            }

            Thread.setDefaultUncaughtExceptionHandler(CrashExceptionHandler())

            // Hook output for logger.
            System.setErr(PrintStream(CustomOutputStream("STDERR", Level.ERROR)))
            System.setOut(PrintStream(CustomOutputStream("STDOUT", Level.INFO)))

            // Logs directory creation.
            References.LOGS_DIR.mkdirs()

            try{
                val fontManager = FontManager()
                loadFonts(fontManager)
            } catch (e: FontFormatException) {
                // Check for null pointer exception.
                if (e is NullPointerException) {
                    System.err.println("Couldn't load fonts.")
                }
                e.printStackTrace()
            } catch (e: NullPointerException) {
                System.err.println("Couldn't load fonts.")
                e.printStackTrace()
            }

            // Create the graphics configurator.
            graphicsConfigurator = GraphicsConfigurator()
            if (!BubbleBlaster.isHeadless()) {
                // Internal canvas.
                window = Window(this, fullscreen, canvas)
                window.cursor = defaultCursor

                // Request focus

                screenManager.onStart { window.isVisible = true }
                canvas.requestFocus()
            }

            // Set font.
            canvas.font = defaultFont

            // Register event bus.
            eventBus.register(this)

            // Start scene-manager.
            Objects.requireNonNull(this.screenManager).start()
        } catch (t: Throwable) {
            // Create crash report instance, and convert to string.
            val crashReport = CrashReport(t)
            val crashString = crashReport.toString()

            // Split the crash report into lines.
            val strings = StringUtils.splitIntoLines(crashString)

            // Print each lien in the stderr output.
            for (string in strings) {
                System.err.println(string)
            }

            // Exit the process with status 1.
            exitProcess(1)
        }
    }
    abstract fun initManagers()

    abstract fun render(gp: GraphicsProcessor)

    abstract fun tick()

    abstract fun loadCursors(manager: CursorManager)

    abstract fun loadFonts(manager: FontManager)
    fun displayScene(scene: Screen?) {
        screenManager.displayScreen(scene)
    }

    fun displayScene(scene: Screen?, force: Boolean) {
        screenManager.displayScreen(scene, force)
    }

    @Suppress("unused")
    fun runLater(runnable: Runnable) {
        tasks.add(runnable)
    }

    protected fun mainThread() {
        val tickCap = 1.0 / BubbleBlaster.TPS.toDouble()
        var frameTime = 0.0
        var frames = 0.0
        var time = Timer.time
        var unprocessed = 0.0
        try {
            while (running) {
                var canTick = false
                val time2 = Timer.time
                val passed = time2 - time
                unprocessed += passed
                frameTime += passed
                time = time2
                while (unprocessed >= tickCap) {
                    unprocessed -= tickCap
                    canTick = true
                }
                if (canTick) {
                    try {
                        tick()
                    } catch (t: Throwable) {
                        val crashReport = CrashReport(t).also {
                            it.addCategory(CrashCategory("Game being ticked").also { it1 ->
                                it1.add("Can tick") { canTick }
                                it1.add("Unprocessed") { unprocessed }
                                it1.add("Tick Cap") { tickCap }
                            })
                        }
                        throw crashReport.reportedException
                    }
                }
                if (frameTime >= 1.0) {
                    frameTime = 0.0
                    fps = frames.roundToLong().toInt()
                    frames = 0.0
                }
                frames++
                try {
                    render()
                } catch (t: Throwable) {
                    val crashReport = CrashReport(t)
                    crashReport.addCategory(CrashCategory("Game being rendered").also{
                        it.add("Frame time") { frameTime }
                        it.add("Frames") { frames }
                        it.add("Fps") { fps }
                    })
                    throw crashReport.reportedException
                }
                for (task in tasks) {
                    task.run()
                }
                tasks.clear()
            }
        } catch (e: ReportedException) {
            if (currentScreen is CrashScreen) {
                val report = CrashReport(e)
                println("Error occurred in displaying crash screen:")
                println(report.toString())
                close(1)
            }

            val crashReport = CrashReport(e)
            val crashString = crashReport.toString()
            println(crashString)
            try {
                displayScene(CrashScreen(e.crashReport), true)
            } catch(t: Throwable) {
                val report = CrashReport(t)
                println("Error occurred in showing crash screen:")
                println(report.toString())
            }

            mainThread()
        }

        // Force shutdown because of crash.
        close(0)
    }

    fun render() {
        val g: Graphics
        var bs: BufferStrategy? = null
        if (BubbleBlaster.isHeadless()) {
            g = NullGraphics()
        } else {
            // Buffer strategy (triple buffering).
            bs = this.canvas.bufferStrategy

            // Create buffers if not created yet.
            if (bs == null) {
                this.canvas.createBufferStrategy(2)
                return
            }

            // Get GraphicsProcessor and GraphicsProcessor objects.
            g = bs.drawGraphics
        }
        val gg = GraphicsProcessor(g)
        if (this.graphicsConfigurator.isAntialiasingEnabled && GameSettings.instance().isTextAntialiasEnabled) gg.hint(
            RenderingHints.KEY_TEXT_ANTIALIASING,
            RenderingHints.VALUE_TEXT_ANTIALIAS_ON
        )
        if (graphicsConfigurator.isAntialiasingEnabled && GameSettings.instance().isAntialiasEnabled) gg.hint(
            RenderingHints.KEY_ANTIALIASING,
            RenderingHints.VALUE_ANTIALIAS_ON
        )
        render(gg)

        currentScreen?.render(this, gg)
        if (this.screenManager.currentScreen != null) screenManager.currentScreen!!.renderGUI(this, gg)

        // Disable Antialias
        if (this.graphicsConfigurator.isAntialiasingEnabled && GameSettings.instance().isTextAntialiasEnabled) gg.hint(
            RenderingHints.KEY_TEXT_ANTIALIASING,
            RenderingHints.VALUE_TEXT_ANTIALIAS_OFF
        )
        if (this.graphicsConfigurator.isAntialiasingEnabled && GameSettings.instance().isAntialiasEnabled) gg.hint(
            RenderingHints.KEY_ANTIALIASING,
            RenderingHints.VALUE_ANTIALIAS_OFF
        )

        val filterEvent = FilterEvent()
        eventBus.post(filterEvent)
        for (filter in filterEvent.filters) {
            gg.composite(FilterComposite(AlphaComposite.getInstance(AlphaComposite.SRC_OVER, 1.0f), filter))
        }

        postRender(GraphicsProcessor(g))

        // Dispose and show.
        g.dispose()
        if (!BubbleBlaster.isHeadless()) {
            bs!!.show()
        }

        // Add key listener.
        logger.info("Loading input event listeners / threads...")
        canvas.addKeyListener(KeyboardController.instance())
        canvas.addMouseListener(MouseController.instance())
        canvas.addMouseMotionListener(MouseController.instance())
        canvas.addMouseWheelListener(MouseController.instance())
    }

    open fun postRender(gp: GraphicsProcessor) {

    }

    open fun start(window: Window) {
        if (canvas.bounds.minX > canvas.bounds.maxX || canvas.bounds.minY > canvas.bounds.maxY) {
            canvas.bounds = window.bounds
        }
        if (canvas.bounds.minX == canvas.bounds.maxX || canvas.bounds.minY == canvas.bounds.maxY) {
            canvas.bounds = window.bounds
        }
        check(!(canvas.bounds.minX > canvas.bounds.maxX || canvas.bounds.minY > canvas.bounds.maxY)) { "Game bounds is invalid: negative size" }
        check(!(canvas.bounds.minX == canvas.bounds.maxX || canvas.bounds.minY == canvas.bounds.maxY)) { "Game bounds is invalid: zero size" }

        running = true

        logger.debug(this.canvas.isVisible)

        mainThread = Thread({ mainThread() }, "TickThread")
        mainThread.start()
    }

    fun configureWindow(window: Window) {
        window.cursor = defaultCursor
    }

    /**
     * Stops the game.
     */
    @Synchronized
    fun stop() {
        try {
            mainThread.join()
            running = false
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     * Lazy shutdown.
     */
    fun shutdown() {
        running = false
    }

    /**
     * Force shutdown the game.
     */
    fun close(code: Int) {
        try {
            // Shut-down game.
            BubbleBlaster.logger.warn("Bubble Blaster 2 is now shutting down...")

            // Check for exit events.
            cleanExit()

            // Stop thread, when fails: interrupt thread, when that also fails show error.
            try {
                // Interrupt thread.
                if (mainThread.isAlive && Thread.currentThread() !== mainThread) {
                    mainThread.interrupt()
                    BubbleBlaster.logger.info("Interrupted main thread.")
                }
            } catch (ignore: SecurityException) {
                BubbleBlaster.logger.error("Cannot stop and interrupt game-thread because of security exceptions.")
            }
        } catch (t: Throwable) {
            try {
                BubbleBlaster.logger.fatal("Exception trying to close Bubble Blaster", t)
            } catch (t: Throwable) {
                exitProcess(1)
            }
            exitProcess(1)
        }
        exitProcess(code)
    }

    /**
     * Force shutdown the game.
     */
    fun close() {
        try {
            // Shut-down game.
            BubbleBlaster.logger.warn("Bubble Blaster 2 is now shutting down...")

            // Check for exit events.
            cleanExit()

            // Stop thread, when fails: interrupt thread, when that also fails show error.
            try {
                // Interrupt thread.
                if (mainThread.isAlive && Thread.currentThread() !== mainThread) {
                    mainThread.interrupt()
                    BubbleBlaster.logger.info("Interrupted main thread.")
                }
            } catch (ignore: SecurityException) {
                BubbleBlaster.logger.error("Cannot stop and interrupt game-thread because of security exceptions.")
            }
        } catch (t: Throwable) {
            try {
                BubbleBlaster.logger.fatal("Exception trying to close Bubble Blaster", t)
            } catch (t: Throwable) {
            }
        }
    }

    open fun cleanExit() {
        eventBus.post(ExitEvent())
    }

    fun requestWindowUserAttention() {
        if (SystemUtils.IS_JAVA_9) {
            // Let the taskbar icon flash. (Java 9+)
            val taskbar = Taskbar.getTaskbar()
            try {
                taskbar.requestWindowUserAttention(window)
            } catch (ignored: UnsupportedOperationException) {
            }
        }
    }

    companion object {
        val logger: Logger = LogManager.getLogger("PreGame")
        var fps = 0
            private set
        lateinit var instance: Game
            private set
    }
}